### ISSUE TITLE
Big test run in stages, yet reports as individual tests

### ISSUE THREAD

**Issue by SDESkowronski** [cypress-io/cypress#3590]
### Current behavior:
Currently when you run a spec file with multiple test cases, they will all run and report their results of either PASS or FAIL.

After one test finished, the state of that test is left behind. It is quite possible for the next test to continue where the prior test left off. This is all good when the prior test passes, but it is not good when it failed since we don't need to waste time running the subsequent tests because we know they will fail. Which is why the Cypress documentation recommends writing bigger test cases that test more features.

### Desired behavior:

We need some mechanism to NOT run a test when a previous test it depends on fails, and to report it as not having been run.
 
I know the docs for Cypress discourages this practice, as have other test teams using other frameworks I have been on. I myself resisted it for a long time. However my manager insisted that I give it a try and when I did I discovered how elegant it actually was.

Prior to trying this I made bigger test cases that tested many features as it drilled down into the main feature being tested. But when I tried breaking those bigger tests into smaller test "stages" as I called them, putting each stage into the `it()` call, I could much more easily see in the test runner UI how the test progressed and at what stage the test failed. Since each stage is comprises of around 3-7 steps, it was much easier to pinpoint exactly where the failure was and in what iteration of the stage if it had been repeated.

Another benefit of this is that these tests became much more self-documenting.

### Versions
Cypress v3.0.3
Chrome Browser v72.0.3626.109

**Comment by jennifer-shehane** [comment]
Hey @SDESkowronski, could you provide some of your test code or psuedo code of an example of how you would like Cypress to run? It would be helpful to understand your use case a bit better.

**Comment by SDESkowronski** [comment]
The pseudo code is below. The important point to note is that after each `it()` block of code, the next `it()` block of code does no setup, it continues where the previous `it()` block left off. If you run this the output in the browser is nice, you can see each stage of the test listed out, and if one of them had failed, then you can quickly see which stages were successful and the first point of failure.

`
describe('Tags and Description', () => {
  context('Train Dialogs', () => {
    before(() => {
      // import the saved model for tags and description testing
    })

    context('Create', () => {
      it('should have no tags are description when creating new dialog', () => {
        // Create new train dialog
        // Verify that tags and description are empty
        expect(true).to.equal(true)
      })

      it('should save the tags and description on the new dialog', () => {
        // Set tags
        // Set description
        // Save
        // Verify tags and description in list
        expect(true).to.equal(true)
      })
    })

    context('Edit', () => {
      it('should open with the tags and description', () => {
        // Find train dialog in list
        // Note the tags and description
        // Open it
        // Verify tags and description are the same as shown in the list
        expect(true).to.equal(true)
      })

      it('should discard the changes made to tags and description when abandoned', () => {
        // Open train dialog
        // Save current tags and description
        // Edit tags
        // Edit description
        // Abandon changes
        // Re-open dialog
        // Verify tags and description are unmodified
        expect(true).to.equal(true)
      })

      it('should save the edited tags and description', () => {
        // Open train dialog
        // Edit tags
        // Edit description
        // Note tags and description
        // reload
        // Verify edited tags and description are still on dialog
        expect(true).to.equal(true)
      })

      it('(advanced edit) should save the edited tags, description, and rounds', () => {
        // Open train dialog
        // Edit tags
        // Edit description
        // Modify dialog to add user input
        // Modify dialog to add bot response
        // Save
        // Reload
        // Re-open dialog
        // Verify edited tags, description, and rounds
        expect(true).to.equal(true)
      })
    })


    context('Continue', () => {
      it('should preserve tags and description when continuing a dialog', () => {
        // Open train dialog
        // Edit the tags
        // Edit the description
        // Type into webchat to continue the dialog
        // Verify the edited tags and description are the same
        // Save the dialog
        // Verify the edited tags and description are in the list
        expect(true).to.equal(true)
      })
    })

    context('Branch', () => {
      it('should preserve tags and description after branching', () => {
        // Open train dialog
        // Edit tags
        // Edit description
        // Click branch on one of the user inputs
        // Enter a new input
        // (Dialog is branched after input is entered)
        // Verify edited tags and description are preserved
        // Save dialog
        // Reload
        // Verify old dialog with original tags and description is still in the list
        // Verify new dialog with edited tags and description is now in the list
        expect(true).to.equal(true)
      })
    })
  })

  context('Log Dialogs', () => {
    before(() => {
      // import model for testing tags and description on log dialogs
    })

    it('should not show tags or description fields on log dialogs', () => {
      // Open log dialog
      // Verify that the tags and description fields do not show up
      // Enter into webchat to continue the dialog
      // Verify that the tags and description fields still do not show up on new window
      expect(true).to.equal(true)
    })
  })
})
`

**Comment by ayandebbarman** [comment]
commenting coz i need this feature .. my scenario is i am validating email once sign up is complete. email validation is its own test case but i don't want it to run if sign up throws error

**Comment by jeanlescure** [comment]
Our team is playing around with Cypress, I love so many of the tool's features, but it worries me that this feature (which can be critical) has been in proposal stage for almost a year 🤔

**Comment by jeanlescure** [comment]
Just to give a little context on how this feature can be critical, one example is infrastructure cost saving. 

The current test setup we have uses AWS CodeBuild to run our battery of automated tests with Cypress. AWS bills us by every 100ms the CodeBuild instance runs, you can then infer that in the long run the costs of using Cypress will add up if when it fails on one test which every other test depends on (i.e. login) it still goes ahead and runs to the end.

**Comment by tamoore** [comment]
I agree with @jeanlescure here. We've also been negotiating purchasing a forest plan with the possibility of going Enterprise. We're currently on hold for all SaaS products because of the recent development in the global economy. However, I think a bail feature is something that should get picked up. I can't seem to find how to achieve this, if it's possible through the even system, would you mind point me to it. I know you can add it per spec, but that seems error prone to me, as one spec might miss it.

**Comment by jennifer-shehane** [comment]
Just wanted to link to issue about option to bail on any failure (different from this, but thought I'd link just in case someone ended up here). https://github.com/cypress-io/cypress/issues/518

**Comment by prafulla-samal** [comment]
i have same kind requirement . i have to reassign an assignment and revome it again. so my test should reassign the assignment then it shoud revoke it. so if reassign fails , i don't want my revoke test to run.

**Comment by talaikis** [comment]
There is simple solution. I am running highly sequential tests that depend on each other and it solves the problem:

```js
afterEach(function () {
    if (this.currentTest.state === 'failed') {
      Cypress.runner.stop()
    }
  })
```

**Comment by quad5** [comment]
You can try this https://github.com/javierbrea/cypress-fail-fast

**Comment by jennifer-shehane** [comment]
As part of Cypress’s [new pricing](https://www.cypress.io/blog/2021/03/16/new-pricing-that-scales-seamlessly/), we’ve included the ability to [cancel test runs when a test fails](https://on.cypress.io/smart-orchestration#Cancel-test-run-when-a-test-fails). This setting is accessible from the Dashboard for organizations starting at the [Business Plan](https://www.cypress.io/pricing/).

This offers a possible solution for the situation outlined in this issue - for those running tests using the Cypress Dashboard - and also ensures a parallelized run is cancelled so that all parallel running specs will also be cancelled to save time on the run.

**To get this feature, you will need to update to Cypress 6.8.0 and also be a member of an organization subscribed to a [Business Plan](https://www.cypress.io/pricing/).**

---

This feature was implemented with parallelized runs in the Dashboard in mind since this was the hardest use case to address. We had to build this feature specifically to continue to receive all of the tests in a cancelled run to ensure proper reporting. Now that we have a mechanism to cancel runs across these channels of communication, we can now consider a way to initiate cancelling test runs when a test fails from the Test Runner when not recording to the Dashboard. (Likely this would be implemented by some CLI flag or config specifying `cancelOnFailures: true)

**This issue https://github.com/cypress-io/cypress/issues/518 will remain open to address the use case of cancelling test runs when a test fails from the Test Runner when not recording to the Dashboard.**

**Comment by cameron** [comment]
Extending @talaikis solution to parallelized test runs (without upgrading to the business plan), consider storing a status flag for a given test run in your (cloud) development db, set it in on failure `afterEach`, and then check it in `beforeEach`.
